transitions/loop.h

```diff
@@ -16,7 +16,39 @@ class TrLoop : public TR {
   bool done() { return false; }
 };
 
-// Usage: TrLoop<PULSE, TRANSITION, OUT>
+// Usage: TrLoopN<N, TRANSITION>
+// N: FUNCTION (number of Loops)
+// TRANSITION: TRANSITION
+// Return Value: TRANSITION
+// Runs the specified transition N times.
+
+template<class N, class TRANSITION>
+class TrLoopN : public TRANSITION {
+public:
+  void run(BladeBase* blade) {
+    n_.run(blade);
+    num_loops_ = n_.calculate(blade);
+    if (loops_ < num_loops_ && TRANSITION::done()) {
+      TRANSITION::begin();
+      loops_ += 1;
+    }
+    TRANSITION::run(blade);
+  }
+  void begin() {
+    TRANSITION::begin();
+    loops_ = 1;
+  }
+  bool done() {
+    return loops_ == num_loops_ && TRANSITION::done(); 
+  }
+
+private:
+  int loops_;
+  int num_loops_;
+  PONUA SVFWrapper<N> n_;
+};
+
+// Usage: TrLoopUntil<PULSE, TRANSITION, OUT>
```

Comment:
I think this can be made smaller and more efficient by counting down, something like:

void run(BladeBase* blade) {
  if (loops_ < 0) loops_ = n_.calculate(blade);
  if (loops_ >  0 && TRANSITION::done()) {
     if (loops_ > 1) TRANSITION::begin();
     loops_--;
  }
  TRANSITION::run(blade);
}
viod begin() {
  TRANSITION::begin();
 loops_ = -1;
}
bool done() { return loops_ == 0; }

This way we don't need "num_loops_", we only call TRANSITION::done() once, and we only call n_.calculate() once per begin call.

Reply:
Thanks, yup much simpler.  Updated and tested.